File: app/src/main/java/org/lineageos/jelly/favorite/FavoriteActivity.java
/*
 * Copyright (C) 2017 The LineageOS Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.lineageos.jelly.favorite;

import android.app.LoaderManager;
import android.content.ContentResolver;
import android.content.ContentUris;
import android.content.CursorLoader;
import android.content.Loader;
import android.database.Cursor;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.appcompat.widget.Toolbar;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.EditText;
import android.widget.LinearLayout;

import org.lineageos.jelly.R;
import org.lineageos.jelly.utils.UiUtils;

public class FavoriteActivity extends AppCompatActivity {
    private RecyclerView mList;
    private View mEmptyView;

    private FavoriteAdapter mAdapter;

    @Override
    protected void onCreate(Bundle savedInstance) {
        super.onCreate(savedInstance);

        setContentView(R.layout.activity_favorites);

        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        toolbar.setNavigationIcon(R.drawable.ic_back);
        toolbar.setNavigationOnClickListener(v -> finish());

        mList = findViewById(R.id.favorite_list);
        mEmptyView = findViewById(R.id.favorite_empty_layout);

        mAdapter = new FavoriteAdapter(this);

        getLoaderManager().initLoader(0, null, new LoaderManager.LoaderCallbacks<Cursor>() {
            @Override
            public Loader<Cursor> onCreateLoader(int id, Bundle args) {
                return new CursorLoader(FavoriteActivity.this, FavoriteProvider.Columns.CONTENT_URI,
                        null, null, null, FavoriteProvider.Columns._ID + " DESC");
            }

            @Override
            public void onLoadFinished(Loader<Cursor> loader, Cursor cursor) {
                mAdapter.swapCursor(cursor);

                if (cursor.getCount() == 0) {
                    mList.setVisibility(View.GONE);
                    mEmptyView.setVisibility(View.VISIBLE);
                }
            }

            @Override
            public void onLoaderReset(Loader<Cursor> loader) {
                mAdapter.swapCursor(null);
            }
        });

        mList.setLayoutManager(new GridLayoutManager(this, 2));
        mList.setItemAnimator(new DefaultItemAnimator());
        mList.setAdapter(mAdapter);

        int listTop = mList.getTop();
        mList.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);

                toolbar.setElevation(recyclerView.getChildAt(0).getTop() < listTop ?
                        UiUtils.dpToPx(getResources(),
                                getResources().getDimension(R.dimen.toolbar_elevation)) : 0);

            }
        });
    }

    void editItem(long id, String title, String url) {
        View view = LayoutInflater.from(this)
                .inflate(R.layout.dialog_favorite_edit, new LinearLayout(this));
        EditText titleEdit = view.findViewById(R.id.favorite_edit_title);
        EditText urlEdit = view.findViewById(R.id.favorite_edit_url);

        titleEdit.setText(title);
        urlEdit.setText(url);

        String error = getString(R.string.favorite_edit_error);
        urlEdit.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (!s.toString().contains("http")) {
                    urlEdit.setError(error);
                }
            }

            @Override
            public void afterTextChanged(Editable s) {
                if (!s.toString().contains("http")) {
                    urlEdit.setError(error);
                }
            }
        });

        new AlertDialog.Builder(this)
                .setTitle(R.string.favorite_edit_dialog_title)
                .setView(view)
                .setPositiveButton(R.string.favorite_edit_positive,
                        ((dialog, which) -> {
                            String updatedUrl = urlEdit.getText().toString();
                            String updatedTitle = titleEdit.getText().toString();
                            if (url.isEmpty()) {
                                urlEdit.setError(error);
                                urlEdit.requestFocus();
                            }
                            new UpdateFavoriteTask(getContentResolver(), id, updatedTitle,
                                    updatedUrl).execute();
                            dialog.dismiss();
                        }))
                .setNeutralButton(R.string.favorite_edit_delete,
                        (dialog, which) -> {
                            new DeleteFavoriteTask(getContentResolver()).execute(id);
                            dialog.dismiss();
                        })
                .setNegativeButton(android.R.string.cancel,
                        (dialog, which) -> dialog.dismiss())
                .show();
    }

    private static class UpdateFavoriteTask extends AsyncTask<Void, Void, Void> {
        private final ContentResolver contentResolver;
        private final long id;
        private final String title;
        private final String url;

        UpdateFavoriteTask(ContentResolver contentResolver, long id, String title, String url) {
            this.contentResolver = contentResolver;
            this.id = id;
            this.title = title;
            this.url = url;
        }

        @Override
        protected Void doInBackground(Void... params) {
            FavoriteProvider.updateItem(contentResolver, id, title, url);
            return null;
        }
    }

    private static class DeleteFavoriteTask extends AsyncTask<Long, Void, Void> {
        private final ContentResolver contentResolver;

        DeleteFavoriteTask(ContentResolver contentResolver) {
            this.contentResolver = contentResolver;
        }

        @Override
        protected Void doInBackground(Long... ids) {
            for (Long id : ids) {
                Uri uri = ContentUris.withAppendedId(FavoriteProvider.Columns.CONTENT_URI, id);
                contentResolver.delete(uri, null, null);
            }
            return null;
        }
    }
}

File: app/src/main/java/org/lineageos/jelly/favorite/FavoriteActivity.kt
/*
 * SPDX-FileCopyrightText: 2020-2024 The LineageOS Project
 * SPDX-License-Identifier: Apache-2.0
 */

package org.lineageos.jelly.favorite

import android.content.ContentResolver
import android.content.ContentUris
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.widget.EditText
import android.widget.LinearLayout
import androidx.activity.viewModels
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.net.toUri
import androidx.core.view.isVisible
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.lineageos.jelly.MainActivity
import org.lineageos.jelly.R
import org.lineageos.jelly.utils.UiUtils
import org.lineageos.jelly.viewmodels.FavoriteViewModel

class FavoriteActivity : AppCompatActivity(R.layout.activity_favorites) {
    // View models
    private val model: FavoriteViewModel by viewModels()

    // Views
    private val favoriteEmptyLayout by lazy { findViewById<View>(R.id.favoriteEmptyLayout) }
    private val favoriteListView by lazy { findViewById<RecyclerView>(R.id.favoriteListView) }
    private val toolbar by lazy { findViewById<Toolbar>(R.id.toolbar) }

    private val adapter by lazy { FavoriteAdapter() }

    override fun onCreate(savedInstance: Bundle?) {
        super.onCreate(savedInstance)

        setSupportActionBar(toolbar)
        supportActionBar?.apply {
            setDisplayHomeAsUpEnabled(true)
            setDisplayShowHomeEnabled(true)
        }

        favoriteListView.layoutManager = GridLayoutManager(this, 2)
        favoriteListView.itemAnimator = DefaultItemAnimator()
        favoriteListView.adapter = adapter
        val listTop = favoriteListView.top
        favoriteListView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                toolbar.elevation = if (recyclerView.getChildAt(0).top < listTop) {
                    UiUtils.dpToPx(resources, resources.getDimension(R.dimen.toolbar_elevation))
                } else {
                    0f
                }
            }
        })

        adapter.onCardClick = { favorite ->
            val intent = Intent(this, MainActivity::class.java).apply {
                data = favorite.url.toUri()
            }
            startActivity(intent)
        }

        adapter.onCardLongClick = { favorite ->
            editItem(favorite.id, favorite.title, favorite.url)
        }

        lifecycleScope.launch {
            lifecycle.repeatOnLifecycle(Lifecycle.State.STARTED) {
                model.favorites.collectLatest {
                    adapter.submitList(it)
                    val empty = it.isEmpty()
                    favoriteListView.isVisible = !empty
                    favoriteEmptyLayout.isVisible = empty
                }
            }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem) = when (item.itemId) {
        android.R.id.home -> {
            finish()
            true
        }

        else -> {
            super.onOptionsItemSelected(item)
        }
    }

    private fun editItem(id: Long, title: String?, url: String) {
        val view = LayoutInflater.from(this)
            .inflate(R.layout.dialog_favorite_edit, LinearLayout(this))
        val titleEdit = view.findViewById<EditText>(R.id.favoriteTitleEditText)
        val urlEdit = view.findViewById<EditText>(R.id.favoriteUrlEditText)
        titleEdit.setText(title)
        urlEdit.setText(url)
        val error = getString(R.string.favorite_edit_error)
        urlEdit.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                if (!s.toString().contains("http")) {
                    urlEdit.error = error
                }
            }

            override fun afterTextChanged(s: Editable) {
                if (!s.toString().contains("http")) {
                    urlEdit.error = error
                }
            }
        })
        AlertDialog.Builder(this)
            .setTitle(R.string.favorite_edit_dialog_title)
            .setView(view)
            .setPositiveButton(
                R.string.favorite_edit_positive
            ) { dialog: DialogInterface, _: Int ->
                val updatedUrl = urlEdit.text.toString()
                val updatedTitle = titleEdit.text.toString()
                if (url.isEmpty()) {
                    urlEdit.error = error
                    urlEdit.requestFocus()
                }
                lifecycleScope.launch {
                    updateFavorite(
                        contentResolver, id, updatedTitle,
                        updatedUrl
                    )
                }
                dialog.dismiss()
            }
            .setNeutralButton(
                R.string.favorite_edit_delete
            ) { dialog: DialogInterface, _: Int ->
                lifecycleScope.launch {
                    deleteFavorite(contentResolver, id)
                }
                dialog.dismiss()
            }
            .setNegativeButton(
                android.R.string.cancel
            ) { dialog: DialogInterface, _: Int -> dialog.dismiss() }
            .show()
    }

    private suspend fun updateFavorite(
        contentResolver: ContentResolver, id: Long,
        title: String, url: String
    ) {
        withContext(Dispatchers.IO) {
            FavoriteProvider.updateItem(contentResolver, id, title, url)
        }
    }

    private suspend fun deleteFavorite(contentResolver: ContentResolver, id: Long) {
        withContext(Dispatchers.IO) {
            val uri = ContentUris.withAppendedId(FavoriteProvider.Columns.CONTENT_URI, id)
            contentResolver.delete(uri, null, null)
        }
    }
}
